### PR TITLE
Add Serialization for RouteManager.

### PR DESCRIPTION
**Pull request by siarheishut**
Add serialization and deserialization for RouteManager.

### CODE REVIEW

File: src/route_manager.cpp [hunk 1]
```diff
@@ -23,6 +59,157 @@ RouteManager::RouteManager(const StopDict &stop_info,
...
+    std::unordered_map<std::string, StopIds> stop_ids) {
+  int vertex_id = 0;
+  for (auto &[stop, _] : stops) {
+    auto &[arrive, depart] = stop_ids[stop];
+    arrive = vertex_id++;
+    depart = vertex_id++;
+  }
+  return std::move(stop_ids);
+}
+
```
Comment (foxfrikses) [inline src/route_manager.cpp:147]: Please try adding these functions and using them. I think this might improve the readability significantly.
```
class StopId {
public:
  explicit StopId(size_t v = 0) : v_(v) {}
  size_t Value() const { return v_; }
private:
  size_t v;
};

class VertexId {
public:
  explicit VertexId(size_t v = 0) : v_(v) {}
  size_t Value() const { return v_; }
private:
  size_t v;
};

bool IsArrive(VertexId vertex);
bool IsDepart(VertexId vertex);
StopId ToStopId(VertexId vertex);
VertexId ToArrive(StopId stopId);
VertexId ToDepart(StopId stopId);
```

File: src/route_manager.cpp [hunk 2]
```diff
@@ -23,6 +59,157 @@ RouteManager::RouteManager(const StopDict &stop_info,
...
+  }
+  for (auto &proto_edge : router.edges()) {
+    edges.emplace_back(RoadEdge{
+        .bus = db.buses().at(proto_edge.bus_id()).bus_name(),
+        .start_idx = static_cast<int>(proto_edge.from_idx()),
+        .span_count = static_cast<int>(
+            proto_edge.to_idx() - proto_edge.from_idx())});
+  }
+  return std::move(edges);
+}
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:202]: Done.

File: src/route_manager.cpp [hunk 3]
```diff
@@ -33,6 +39,76 @@ size_t ToArrive(size_t stop_id) {
...
+                        .from = ToArrive(id),
+                        .to = ToDepart(id),
+                        .weight = static_cast<double>(bus_wait_time_)});
+    }
+    for (auto &proto_edge : router_.edges()) {
+      auto bus = db_.buses().at(proto_edge.bus_id());
+      auto from = bus.route().at(proto_edge.from_idx());
+      auto to = bus.route().at(proto_edge.to_idx());
+      graph.AddEdge({
+                        .from = from * 2 + 1,
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:99]: Done.

File: src/route_manager.cpp [hunk 4]
```diff
@@ -44,9 +131,88 @@ RouteManager::RouteManager(const StopDict &stop_info,
...
+
+  auto stop_indices = IndicesMap(catalog_->Stops());
+  auto bus_indices = IndicesMap(catalog_->Buses());
+
+  for (auto &proto_edge : ProtoEdges(graph_, edges_, bus_indices)) {
+    *proto_router.add_edges() = std::move(proto_edge);
+  }
+
+  vector<string> idx_to_bus;
+  transform(begin(catalog_->Buses()), end(catalog_->Buses()), begin(idx_to_bus),
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:149]: Done.

File: src/route_manager.cpp [hunk 5]
```diff
@@ -123,4 +289,46 @@ std::optional<RouteInfo> RouteManager::FindRoute(const std::string &from,
 
   return route_info;
 }
+
+vector<::TransportCatalog::Edge> RouteManager::ProtoEdges(
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:293]: Done.

File: src/route_manager.cpp [hunk 6]
```diff
@@ -33,6 +40,86 @@ size_t ToArrive(size_t stop_id) {
...
 }
+
+bool ValidateEdges(const ::TransportCatalog::TransportDatabase &db,
+                   const ::TransportCatalog::Router &router) {
+  return all_of(
+      router.edges().begin(), router.edges().end(),
+      [&](auto &edge) {
+        return (edge.bus_id() >= db.buses_size() ||
+            edge.from_idx() >= db.buses().at(edge.bus_id()).route_size() ||
+            edge.to_idx() >= db.buses().at(edge.bus_id()).route_size());
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:51]: Done.

[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:51]: Done.

File: src/route_manager.cpp [hunk 7]
```diff
@@ -33,6 +40,86 @@ size_t ToArrive(size_t stop_id) {
...
+  return routes;
+}
+
+struct GraphBuilder {
+  const ::TransportCatalog::TransportDatabase &db_;
+  const ::TransportCatalog::Router &router_;
+  const unordered_map<string, size_t> &stop_ids_;
+  int bus_wait_time_;
+
+  vector<rm::route_manager::Edge> BuildEdges() const {
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:72]: Done.

File: src/route_manager.cpp [hunk 8]
```diff
@@ -33,6 +40,86 @@ size_t ToArrive(size_t stop_id) {
...
+      edges.emplace_back(WaitEdge{});
+    }
+    for (auto &proto_edge : router_.edges()) {
+      edges.emplace_back(RoadEdge{
+          .bus = db_.buses().at(proto_edge.bus_id()).bus_name(),
+          .start_idx = static_cast<int>(proto_edge.from_idx()),
+          .span_count = static_cast<int>(
+              proto_edge.to_idx() - proto_edge.from_idx())});
+    }
+    return std::move(edges);
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:84]: Done.

File: src/route_manager.cpp [hunk 9]
```diff
@@ -33,6 +40,86 @@ size_t ToArrive(size_t stop_id) {
...
+  }
+  rm::route_manager::Graph BuildGraph() const {
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:86]: Done.

File: src/route_manager.cpp [hunk 10]
```diff
@@ -33,6 +40,86 @@ size_t ToArrive(size_t stop_id) {
...
+    for (auto &proto_edge : router_.edges()) {
+      auto bus = db_.buses().at(proto_edge.bus_id());
+      auto from = bus.route().at(proto_edge.from_idx());
+      auto to = bus.route().at(proto_edge.to_idx());
+      graph.AddEdge({
+                        .from = from * 2 + 1,
+                        .to = to * 2,
+                        .weight = proto_edge.weight()});
+    }
+    return std::move(graph);
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:103]: Done.

File: src/route_manager.cpp [hunk 11]
```diff
@@ -33,6 +40,86 @@ size_t ToArrive(size_t stop_id) {
...
+  }
+};
+
+::TransportCatalog::RoutingSettings InitializeSettings(RoutingSettings s) {
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:107]: Done.

File: src/route_manager.cpp [hunk 12]
```diff
@@ -44,9 +131,88 @@ RouteManager::RouteManager(const StopDict &stop_info,
...
+
+  for (auto &proto_edge : ProtoEdges(graph_, edges_, bus_indices)) {
+    *proto_router.add_edges() = std::move(proto_edge);
+  }
+
+  vector<string> idx_to_bus;
+  transform(begin(catalog_->Buses()), end(catalog_->Buses()), begin(idx_to_bus),
+            [](const auto &p) { return p.first; });
+
+  unordered_map<tuple<int, int, double>, size_t,
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:152]: Done.

File: src/route_manager.cpp [hunk 13]
```diff
@@ -44,9 +131,88 @@ RouteManager::RouteManager(const StopDict &stop_info,
       stop_names_(stop_info.size()) {
   ReadStops(stop_info);
   ReadBuses(bus_info, stop_info);
-  router_ = std::make_unique<Router>(graph_);
+  router_ = make_unique<Router>(graph_);
 }
 
+::TransportCatalog::Router RouteManager::Serialize() const {
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:137]: Done.

File: src/route_manager.cpp [hunk 14]
```diff
@@ -123,4 +289,46 @@ std::optional<RouteInfo> RouteManager::FindRoute(const std::string &from,
...
+      continue;
+    auto &edge = get<RoadEdge>(edges[i]);
+    auto &graph_edge = graph.GetEdge(i);
+
+    ::TransportCatalog::Edge proto_edge;
+    proto_edge.set_weight(graph_edge.weight);
+    proto_edge.set_from_idx(edge.start_idx);
+    proto_edge.set_to_idx(edge.start_idx + edge.span_count);
+    proto_edge.set_bus_id(bus_indices.at(edge.bus));
+    result.push_back(std::move(proto_edge));
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:309]: Done.

File: src/route_manager.cpp [hunk 15]
```diff
@@ -123,4 +289,46 @@ std::optional<RouteInfo> RouteManager::FindRoute(const std::string &from,
...
+  for (auto [from, to, prev_edge] : router.Serialize()) {
+    if (IsDepart(from) || IsDepart(to)) continue;
+
+    ::TransportCatalog::Route proto_route;
+    proto_route.set_from(ToStopId(from));
+    proto_route.set_to(ToStopId(to));
+    auto &[vertex_from, vertex_to, prev_weight] = graph.GetEdge(prev_edge);
+    proto_route.set_last_edge_id(proto_edge_indices[{ToStopId(vertex_from),
+                                                     ToStopId(vertex_to),
+                                                     prev_weight}]);
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:329]: Done.

File: src/route_manager.cpp [hunk 16]
```diff
@@ -33,6 +40,86 @@ size_t ToArrive(size_t stop_id) {
...
+      });
+}
+
+vector<graph::Route> Routes(const ::TransportCatalog::Router &router) {
+  vector<graph::Route> routes;
+  for (auto &proto_route : router.routes()) {
+    auto &route = routes.emplace_back();
+    route.from = proto_route.from();
+    route.to = proto_route.to();
+    route.prev_edge = proto_route.last_edge_id();
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:61]: Done.

File: src/route_manager.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (foxfrikses) [reply src/route_manager.cpp]: The part with `check for from_idx and to_idx not pointing to the same stop` is not done yet.

I would also remove `()` around the logical operations.

File: src/route_manager_utils.h [hunk 17]
```diff
@@ -19,6 +26,30 @@ struct RoadEdge {
 struct WaitEdge {};
 
 using Edge = std::variant<RoadEdge, WaitEdge>;
+
+struct Params {
```
Comment (foxfrikses) [inline src/route_manager_utils.h:30]: It's not intuitive to define a constructor params in a different file from the constructor. Could you move it to the file, where it's actually being used?

File: src/route_manager.cpp [hunk 18]
```diff
@@ -33,6 +40,155 @@ size_t ToArrive(size_t stop_id) {
...
+  rm::route_manager::Graph BuildGraph() const {
+    rm::route_manager::Graph graph(stop_ids_.size() * 2);
+    for (auto &[stop, id] : stop_ids_) {
+      graph.AddEdge({
+                        .from = ToArrive(id),
+                        .to = ToDepart(id),
+                        .weight = static_cast<double>(bus_wait_time_)});
+    }
+    for (auto &proto_edge : router_.edges()) {
+      auto bus = db_.buses().at(proto_edge.bus_id());
```
Comment (foxfrikses) [inline src/route_manager.cpp:102]:
```suggestion
      auto& bus = db_.buses().at(proto_edge.bus_id());
```

File: src/route_manager.cpp [hunk 19]
```diff
@@ -33,6 +40,155 @@ size_t ToArrive(size_t stop_id) {
...
+    proto_edge.set_weight(graph_edge.weight);
+    proto_edge.set_from_idx(edge.start_idx);
+    proto_edge.set_to_idx(edge.start_idx + edge.span_count);
+    proto_edge.set_bus_id(bus_indices.at(edge.bus));
+    result.push_back(std::move(proto_edge));
+  }
+  return result;
+}
+
+// Works due to the fact, that all wait edges are always being maintained before
```
Comment (foxfrikses) [inline src/route_manager.cpp:152]:
```
Based on the knowledge that the first half of edges are wait-edges, and the second half are route-edges.
```

File: src/route_manager.cpp [hunk 20]
```diff
@@ -33,6 +40,155 @@ size_t ToArrive(size_t stop_id) {
...
+  unordered_map<size_t, size_t> result;
+  int edge_id = stop_count, proto_edge_id = 0;
+  while (edge_id < edges.size()) {
+    result[edge_id] = proto_edge_id;
+    ++edge_id, ++proto_edge_id;
+  }
+  return result;
+}
+
+// Works due to the fact, that all wait edges are always being maintained before
```
Comment (foxfrikses) [inline src/route_manager.cpp:165]: Based on the knowledge that the first half of edges are wait-edges, and the second half are route-edges.

File: src/route_manager_utils.h [hunk 21]
```diff
@@ -19,6 +26,30 @@ struct RoadEdge {
...
+  std::unordered_map<std::string, size_t> stop_ids;
+};
+
+struct Route {
+  size_t from;
+  size_t to;
+  size_t bus_id;
+};
+
+struct kHasher {
```
Comment (foxfrikses) [inline src/route_manager_utils.h:46]: You don't need this struct, am I right?

File: src/route_manager_utils.h [hunk 22]
```diff
@@ -19,6 +26,30 @@ struct RoadEdge {
...
+  std::shared_ptr<TransportCatalog> catalog;
+  std::unique_ptr<graph::Router<double>> router;
+  graph::DirectedWeightedGraph<double> graph;
+  utils::RoutingSettings settings;
+  std::vector<route_manager::Edge> edges;
+  std::vector<std::string> stop_names;
+  std::unordered_map<std::string, size_t> stop_ids;
+};
+
+struct Route {
```
Comment (foxfrikses) [inline src/route_manager_utils.h:40]: You don't need this struct, am I right?

File: src/route_manager.cpp [hunk 23]
```diff
@@ -33,6 +40,155 @@ size_t ToArrive(size_t stop_id) {
...
+    route.prev_edge = edge_ids[proto_route.last_edge_id()];
+  }
+  return routes;
+}
+
+struct GraphBuilder {
+  const ::TransportCatalog::TransportDatabase &db_;
+  const ::TransportCatalog::Router &router_;
+  const unordered_map<string, size_t> &stop_ids_;
+  int bus_wait_time_;
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:72]: Done.

File: src/route_manager.cpp [hunk 24]
```diff
@@ -33,6 +40,155 @@ size_t ToArrive(size_t stop_id) {
...
+
+  // All buses in router_.edges are expected to be inside db.buses.
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:74]: Done.

File: src/route_manager_utils.h (hunk 17 above)
Comment (siarheishut) [reply src/route_manager_utils.h:30]: Done.

File: src/route_manager.cpp (hunk 19 above)
Comment (siarheishut) [reply src/route_manager.cpp:152]: Done.

File: src/route_manager.cpp (hunk 20 above)
Comment (siarheishut) [reply src/route_manager.cpp:165]: Done.

File: src/route_manager_utils.h (hunk 21 above)
Comment (siarheishut) [reply src/route_manager_utils.h:46]: You do, removed.

File: src/route_manager.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp]: Done.

File: src/route_manager.cpp [hunk 25]
```diff
@@ -33,6 +40,155 @@ size_t ToArrive(size_t stop_id) {
...
+  return all_of(router.edges().begin(), router.edges().end(), validate_edge);
+}
+
+vector<graph::Route> Routes(const ::TransportCatalog::Router &router,
+                            unordered_map<size_t, size_t> edge_ids) {
+  vector<graph::Route> routes;
+  for (auto &proto_route : router.routes()) {
+    auto &route = routes.emplace_back();
+    route.from = ToArrive(proto_route.from());
+    route.to = ToArrive(proto_route.to());
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/route_manager.cpp:62]: Done.

File: src/route_manager.h [hunk 26]
```diff
@@ -4,31 +4,56 @@
...
 
 #include "common.h"
+#include "graph.h"
 #include "request_types.h"
 #include "route_manager_utils.h"
+#include "transport_catalog.h"
+#include "transport_catalog.pb.h"
 
 namespace rm {
+struct Params {
```
Comment (foxfrikses) [inline src/route_manager.h:23]: Better to the private section of the RouteManager class itself

File: src/route_manager.cpp [hunk 27]
```diff
@@ -33,6 +40,176 @@ size_t ToArrive(size_t stop_id) {
...
+            db.stops().at(edge.to_idx()).stop_name());
+  };
+
+  return all_of(router.edges().begin(), router.edges().end(), validate_edge);
+}
+
+vector<graph::Route> Routes(
+    const ::TransportCatalog::TransportDatabase db,
+    const ::TransportCatalog::Router &router,
+    const unordered_map<std::string, size_t> &stop_ids,
```
Comment (foxfrikses) [inline src/route_manager.cpp:61]: Could you make it size_t -> size_t map. And create a function for preparing this map?

File: src/route_manager.h (hunk 26 above)
Comment (siarheishut) [reply src/route_manager.h:23]: Done.

File: src/route_manager.cpp (hunk 27 above)
Comment (siarheishut) [reply src/route_manager.cpp:61]: Done.

File: src/route_manager_utils.h [hunk 28]
```diff
@@ -1,11 +1,18 @@
...
 #include <string>
+#include <tuple>
+#include <unordered_map>
 #include <variant>
+#include <vector>
 
 #include "graph.h"
+#include "request_types.h"
 #include "router.h"
+#include "transport_catalog.h"
```
Comment (foxfrikses) [inline src/route_manager_utils.h:15]: I don't think you should modify any include in this file in this commit, do you?

File: src/route_manager.cpp [hunk 29]
```diff
@@ -44,9 +223,78 @@ RouteManager::RouteManager(const StopDict &stop_info,
       stop_names_(stop_info.size()) {
   ReadStops(stop_info);
   ReadBuses(bus_info, stop_info);
-  router_ = std::make_unique<Router>(graph_);
+  router_ = make_unique<Router>(graph_);
```
Comment (foxfrikses) [inline src/route_manager.cpp:222]: The fact that you added `using namespace std` in this file doesn't force you to remove std:: from every place here: it misleads when you read the changes. Could you undo these modifications where it's not the function, that you modify/create completely?

File: src/route_manager.cpp [hunk 30]
```diff
@@ -33,6 +40,174 @@ size_t ToArrive(size_t stop_id) {
...
+
+::TransportCatalog::RoutingSettings ToProto(RoutingSettings s) {
+  ::TransportCatalog::RoutingSettings proto_routing_settings;
+  proto_routing_settings.set_bus_wait_time(s.bus_wait_time);
+  proto_routing_settings.set_bus_velocity(s.bus_velocity);
+  return proto_routing_settings;
+}
+
+template<typename Container>
+unordered_map<string_view, size_t> BusIndices(Container cont) {
```
Comment (foxfrikses) [inline src/route_manager.cpp:128]: `const Container&`, otherwise segfault

File: src/route_manager.cpp [hunk 31]
```diff
@@ -33,6 +40,174 @@ size_t ToArrive(size_t stop_id) {
...
+  while (edge_id < edges.size()) {
+    result[edge_id] = proto_edge_id;
+    ++edge_id, ++proto_edge_id;
+  }
+  return result;
+}
+
+// Based on the knowledge that the first half of edges are wait-edges, and the
+// second half are route-edges.
+unordered_map<size_t, size_t> EdgeIds(
```
Comment (foxfrikses) [inline src/route_manager.cpp:173]: Let's name `ProtoEdgeIdToEdgeId` and `EdgeIdToProtoEdgeId` for consistency with stops

File: lib/graph/src/router.h [hunk 32]
```diff
@@ -208,7 +208,7 @@ std::unique_ptr<Router<Weight>> Router<Weight>::Deserialize(
     st.emplace(from, to);
     st.emplace(from, second_to_last);
   }
-  return std::make_unique<Router>(graph, routes_internal_data);
+  return std::make_unique<Router>(new Router(graph, routes_internal_data));
```
Comment (foxfrikses) [inline lib/graph/src/router.h:211]: Could you put this is a separate commit?

File: src/route_manager.cpp (hunk 30 above)
Comment (siarheishut) [reply src/route_manager.cpp:128]: Done.

File: src/route_manager.cpp (hunk 29 above)
Comment (siarheishut) [reply src/route_manager.cpp:222]: Removed.

File: src/route_manager_utils.h (hunk 28 above)
Comment (siarheishut) [reply src/route_manager_utils.h:15]: I definitely don't. Removed.

File: lib/graph/src/router.h (hunk 32 above)
Comment (siarheishut) [reply lib/graph/src/router.h:211]: Done.

File: src/route_manager.cpp (hunk 31 above)
Comment (siarheishut) [reply src/route_manager.cpp]: Done.

File: src/route_manager_utils.h [hunk 33]
```diff
@@ -2,7 +2,7 @@
 #define ROOT_MANAGER_SRC_ROUTE_MANAGER_UTILS_H_
 
 #include <string>
-#include <variant>
+#include <tuple>
```
Comment (foxfrikses) [inline src/route_manager_utils.h:5]:
```suggestion
#include <variant>
```

File: src/route_manager.cpp [hunk 34]
```diff
@@ -33,6 +40,174 @@ size_t ToArrive(size_t stop_id) {
...
+
+std::unordered_map<size_t, size_t> ProtoStopIdToStopId(
+    const unordered_map<string, size_t> &stop_ids,
+    const ::TransportCatalog::TransportDatabase &db) {
+  std::unordered_map<size_t, size_t> result;
+  for (int i = 0; i < db.stops().size(); ++i) {
+    auto &name = db.stops().at(i).stop_name();
+    if (stop_ids.find(name) != stop_ids.end()) {
+      result[i] = stop_ids.at(name);
+    }
```
Comment (foxfrikses) [inline src/route_manager.cpp:192]: else log that the name was not found in stop_ids

File: lib/graph/src/router.h [hunk 35]
```diff
@@ -208,7 +208,7 @@ std::unique_ptr<Router<Weight>> Router<Weight>::Deserialize(
     st.emplace(from, to);
     st.emplace(from, second_to_last);
   }
-  return std::make_unique<Router>(graph, routes_internal_data);
+  return std::unique_ptr<Router>(new Router(graph, routes_internal_data));
```
[the remark this answers is not in the document]
Comment (foxfrikses) [reply lib/graph/src/router.h:211]: After you merge https://github.com/siarheishut/route_manager/pull/121, pull the changes to this commit.

File: src/route_manager_utils.h (hunk 33 above)
Comment (siarheishut) [reply src/route_manager_utils.h:5]: Done.

File: src/route_manager.cpp (hunk 34 above)
Comment (siarheishut) [reply src/route_manager.cpp:192]: Done.

File: lib/graph/src/router.h (hunk 35 above)
[the remark this answers is not in the document]
Comment (siarheishut) [reply lib/graph/src/router.h:211]: Done.
